lib/fsm/follow.go

```diff
@@ -59,66 +61,82 @@ func (e *PlanRolledBackEvent) isTerminalEvent() bool { return true }
 
 // FollowOperationPlan returns a channel that will be receiving phase updates
 // for the specified plan.
+// The returned channel must be served to avoid blocking the producer as
+// no events are dropped.
 //
 // The watch will stop upon entering one of the terminal operation states, for
-// example if the obtained plan is completed or fully rolled back.
+// example if the obtained plan is completed or fully rolled back or when the specified
+// context has expired. In any case, the channel will be closed to signal completion
 func FollowOperationPlan(ctx context.Context, getPlan GetPlanFunc) <-chan PlanEvent {
-	ch := make(chan PlanEvent, 100)
-	// Send an initial batch of events from the initial state of the plan.
-	plan, err := getPlan()
-	if err != nil {
-		logrus.WithError(err).Error("Failed to load plan.")
-	}
-	if plan != nil {
-		events := getPlanEvents(GetPlanProgress(*plan), *plan)
-		// recreate channel with enough room for all events + some room for additional events
-		ch = make(chan PlanEvent, len(events)+10)
-
-		for _, event := range events {
-			select {
-			case ch <- event:
-			default:
-				logrus.WithField("event", event).Warn("Event channel is full.")
+	ch := make(chan PlanEvent)
+	go func() {
+		defer close(ch)
+		// Send an initial batch of events from the initial state of the plan.
+		plan, err := getPlan()
+		if err == nil {
+			err = sendPlanChanges(ctx, GetPlanProgress(*plan), *plan, ch)
+			if errors.Is(err, errStopUpdates) {
+				// Done
+				return
+			}
+		}
+		ticker := backoff.NewTicker(getFollowStepPolicy())
+		tickerC := ticker.C
+		var errorTicker *backoff.Ticker
+		defer func() {
+			if ticker != nil {
+				ticker.Stop()
+			}
+			if errorTicker != nil {
+				errorTicker.Stop()
 			}
-			if event.isTerminalEvent() {
-				return ch
+		}()
+		resetBackoff := func() {
+			if errorTicker != nil {
+				errorTicker.Stop()
+				errorTicker = nil
 			}
+			ticker = backoff.NewTicker(getFollowStepPolicy())
+			tickerC = ticker.C
+		}
+		startBackoff := func() {
```

Comment:
I sort of wonder if swapping the timers complexity is really required. What I mean is will this follower really reduce enough load by backing off to make a different on transient errors? And if it does, could the normal ticker just be a little bit slower?

Reply:
good question. I can live with a single ticker and update the step to 3-5s - wdyt?

Reply:
Yea, while I don't see a problem with the code as implemented, I think 3-5s is fine for the behaviour and probably alot simpler.